src/encoding/error.rs

```diff
@@ -1,28 +1,36 @@
 #[cfg(feature = "std")]
 use std::sync::Arc;
 
-use failure::Fail;
+use snafu::Snafu;
 
-use crate::state_tracker::StructureError;
+use crate::state_tracker;
 
-#[derive(Debug, Clone, Fail)]
-#[fail(display = "encoding failed: {}", _0)]
-pub struct Error(#[fail(cause)] pub ErrorKind);
+#[derive(Debug, Clone, Snafu)]
+#[snafu(display("encoding failed: {}", source))]
+pub struct Error {
```

Comment:
I think it's unrelated to this PR, but is there a reason that there are separate `Error` and `ErrorKind` types? Is it for future flexibility in case you want to add things to the `Error` type?

Reply:
As far as I remember the idea was to come up with an error type that allows us to add additional variants without any concerns about potential compatibility issues. I know this sometimes increases the workload to actualy handle a specifc error. 

Do you think its worth to discuss this decision?

Reply:
I usually see this pattern used when you have an type with many variants, but there are fields with are common to all variants, so instead of doing:

```rust
enum Foo {
  A { id: u32 },
  B { id: u32 },
  C { id: u32 },
}
```

You do:

```rust

struct Foo {
  id: u32,
  kind: FooKind,
}

enum FooKind {
  A,
  B,
  C,
}
```

It stood out to me in this case because doesn't seem to be any shared data between variants. I don't think it's done to allow backwards compatible extension with new variants, usually that's done by marking the enum as `#[non_exhaustive]`.

But, if you might add shared data between variants, or if you don't mind a breaking change in the future, then I think this isn't critical to address.

Reply:
I can see how `#[non_exhaustive]` would allow to simplify everything. I could rewrite the structure but I assume thats something @thequux should decide.

Reply:
Yeah, I think that `#[non_exhaustive]` makes more sense here than the ErrorKind structure. IIRC, we used the ErrorKind split because it was recommended in https://rust-lang-nursery.github.io/failure/error-errorkind.html, and we didn't think too closely about it at the time.